samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdlib.h>
+
+#include <zephyr.h>
+#include <settings/settings.h>
+
+#include "event_manager.h"
+#include "config_event.h"
+
+#define MODULE config
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CONFIG_LOG_LEVEL);
+
+#define KEY_LEN 9
+
+/* Array of pointers to data loaded from settings */
+static u8_t *loaded_data[CONFIG_EVENT_ID_MAX];
+
+static bool config_id_is_supported(u8_t id)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE)) {
+		if ((id >= CONFIG_EVENT_ID_MOTION_CPI) &&
+		    (id <= CONFIG_EVENT_ID_MOTION_DOWNSHIFT_REST2)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int settings_set(int argc, char **argv, void *val_ctx)
+{
+	if (argc != 1) {
+		return -ENOENT;
+	}
+
+	char *end;
+	u8_t id = strtol(argv[0], &end, 10);
+	if (*end != '\0') {
+		LOG_WRN("Loaded config ID is not a valid number");
+		return 0;
+	}
+
+	if (!config_id_is_supported(id)) {
+		LOG_WRN("Config ID not supported in current configuration: %u", id);
+		return 0;
+	}
+
+	if (!loaded_data[id]) {
+		loaded_data[id] = k_malloc(REPORT_SIZE_USER_CONFIG - 1);
+
+		if (loaded_data[id] == NULL) {
+			LOG_ERR("No memory");
+			module_set_state(MODULE_STATE_ERROR);
+			return -ENOMEM;
+		}
+	}
+
+	int len = settings_val_read_cb(val_ctx, loaded_data[id],
+				       REPORT_SIZE_USER_CONFIG - 1);
+	if (len < 0) {
+		LOG_ERR("Can't read config (id:%u err:%d)", id, len);
+		return len;
+	}
+
+	return 0;
+}
+
+static int commit(void)
+{
+	/* All settings loaded, including FCB duplicates.
+	 * We can apply values to application.
+	 */
+	for (u8_t id = 0; id < CONFIG_EVENT_ID_MAX; id++) {
+		if (!loaded_data[id]) {
+			continue;
+		}
+
+		__ASSERT_NO_MSG(id != CONFIG_EVENT_ID_RESERVED);
+
+		struct config_event *event = new_config_event();
+
+		event->id = id;
+		event->stored = true;
+		memcpy(event->data, loaded_data[id], sizeof(event->data));
+
+		EVENT_SUBMIT(event);
+
+		k_free(loaded_data[id]);
+	}
+
+	return 0;
+}
+
+static int update_config(const u8_t config_id, u8_t *data)
+{
+	int err;
+
+	char key[KEY_LEN];
+	err = snprintk(key, sizeof(key), MODULE_NAME "/%u", config_id);
+	if (err < 0) {
+		LOG_ERR("Creating setting key failed");
+		return err;
+	}
+
```

Comment:
`ASSERT(err < sizeof(key))` to avoid problems when module name changes

Reply:
done